src/platform_impl/linux/system_tray.rs

```diff
@@ -41,43 +32,109 @@ impl SystemTrayBuilder {
   ) -> Result<RootSystemTray, OsError> {
     let sender = window_target.p.window_requests_tx.clone();
 
-    if let Some(tray_menu) = self.tray_menu.clone() {
-      let menu = &mut tray_menu.into_gtkmenu(&sender, &AccelGroup::new(), WindowId::dummy());
-
-      self.app_indicator.set_menu(menu);
-      menu.show_all();
-    }
-
-    self.app_indicator.set_status(AppIndicatorStatus::Active);
-
-    Ok(RootSystemTray(SystemTray {
-      app_indicator: self.app_indicator,
-      sender,
-    }))
+    Ok(RootSystemTray(SystemTray::new(self.tray, sender)))
   }
 }
 
 pub struct SystemTray {
-  app_indicator: AppIndicator,
+  tray_handle: ksni::Handle<KsniTray>,
   sender: Sender<(WindowId, WindowRequest)>,
 }
 
 impl SystemTray {
+  pub fn new(tray: KsniTray, sender: Sender<(WindowId, WindowRequest)>) -> Self {
+    let tray_service = ksni::TrayService::new(tray);
+    let tray_handle = tray_service.handle();
+    tray_service.spawn();
+
+    Self {
+      tray_handle: tray_handle,
+      sender,
+    }
+  }
+
   pub fn set_icon(&mut self, icon: PathBuf) {
-    let path = icon.parent().expect("Invalid icon");
-    self
-      .app_indicator
-      .set_icon_theme_path(&path.to_string_lossy());
-    self.app_indicator.set_icon(&icon.to_string_lossy())
+    self.tray_handle.update(|tray: &mut KsniTray| {
+      tray.set_icon(&icon);
+    });
   }
 
   pub fn set_menu(&mut self, tray_menu: &Menu) {
-    let mut menu =
-      tray_menu
-        .clone()
-        .into_gtkmenu(&self.sender, &AccelGroup::new(), WindowId::dummy());
+    //let mut menu =
+    //  tray_menu
+    //    .clone()
+    //    .into_gtkmenu(&self.sender, &AccelGroup::new(), WindowId::dummy());
+
+    //self.app_indicator.set_menu(&mut menu);
+    //menu.show_all();
+  }
+}
+
+/// Holds all properties and signals of the tray and manages the communcation via DBus.
+pub struct KsniTray {
+  title: String,
+  icon_name: String,
+  icon_theme_path: String,
+  status: ksni::Status,
+}
+
+impl KsniTray {
+  /// Initializes a new instance.
+  ///
+  /// # Arguments
+  ///
+  /// * `title` -  The instance title.
+  /// * `icon` -  Absolute file path to the icon that will be visible in tray.
+  ///
+  /// Initial status is set to `ksni::Status::Active`
+  pub fn new(title: &str, icon: &PathBuf) -> Self {
+    let (icon_name, icon_theme_path) = Self::split_icon(&icon);
+
+    Self {
+      title: title.to_string(),
+      icon_name,
+      icon_theme_path,
+      status: ksni::Status::Active,
+    }
+  }
+
+  /// Updates the icon.
+  pub fn set_icon(&mut self, icon: &PathBuf) {
+    let (icon_name, icon_theme_path) = Self::split_icon(&icon);
+    self.icon_name = icon_name;
+    self.icon_theme_path = icon_theme_path;
+  }
+
+  fn split_icon(icon: &PathBuf) -> (String, String) {
```

Comment:
For the icon could we use a Vec<u8> like other platforms? It could be really good if we can remove that `PathBuf`

Reply:
A method `icon_pixmap` exists that holds the icon in binary representation, so it should be possible. It requires an ARGB32 representation of the image.
https://docs.rs/ksni/0.1.3/ksni/trait.Tray.html#method.icon_pixmap